src/Database/LSMTree/Model/Monoidal.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE RoleAnnotations          #-}
+{-# LANGUAGE ScopedTypeVariables      #-}
+{-# LANGUAGE StandaloneDeriving       #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TupleSections            #-}
+{-# LANGUAGE TypeApplications         #-}
+{-# LANGUAGE TypeFamilies             #-}
+
+-- lookup has redundant update constraint.
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+-- |
+--
+-- This module is intended to be imported qualified.
+--
+-- > import qualified Database.LSMTree.Monoidal as LSMT
+module Database.LSMTree.Model.Monoidal (
+    -- * Temporary placeholder types
+    SomeSerialisationConstraint (..)
+  , SomeUpdateConstraint (..)
+    -- * Tables
+  , Table
+  , empty
+    -- * Table querying and updates
+    -- ** Queries
+  , Range
+  , LookupResult (..)
+  , lookups
+  , RangeLookupResult (..)
+  , rangeLookup
+    -- ** Updates
+  , Update (..)
+  , updates
+  , inserts
+  , deletes
+  , mupserts
+    -- * Snapshots
+  , snapshot
+    -- * Multiple writable table handles
+  , duplicate
+    -- * Merging tables
+  , mergeTables
```

Comment:
This reminds me that we named it wrong in the API module. We decided on not including "table" in all the names. So this should have been just called `union`. Anyway, that TODO is for another PR.